### PR TITLE
Drop update-ref todos pointing to head

### PR DESCRIPTION
**Pull request by stefanhaller**
- **PR Description**

The rebase.updateRefs feature of git is very useful to rebase a stack of branches and keep everything nicely stacked; however, it is usually in the way when you make a copy of a branch and want to rebase it "away" from the original branch in some way or other. For example, the original branch might sit on main, and you want to rebase the copy onto devel to see if things still compile there. Or you want to do some heavy history rewriting experiments on the copy, but keep the original branch in case the experiments fail. Or you want to split a branch in two because it contains two unrelated sets of changes; so you make a copy, and drop half of the commits from the copy, then check out the original branch and drop the other half of the commits from it.

In all these cases, git's updateRefs feature insists on moving the original branch along with the copy in the first rebase that you make on the copy. I think this is a bug in git, it should create update-ref todos only for branches that point into the middle of your branch (because only then do they form a stack), not when they point at the head (because then it's a copy). I had a long discussion about this on the git mailing list [1], but people either don't agree or don't care enough.

So we fix this on our side: whenever we start a rebase for whatever reason, be it interactive, non-interactive, or behind-the-scenes, we drop any update-ref todos that are at the very top of the todo list, which fixes all the above-mentioned scenarios nicely.

I will admit that there's one scenario where git's behavior is the desired one, and the fix in this PR makes it worse: when you create a new branch off of an existing one, with the intention of creating a stack of branches, but before you make the first commit on the new branch you realize some problem with the first branch (e.g. a commit that needs to be reworded or dropped). It this case you do want both branches to be affected by the change. In my experience this scenario is much rarer than the other ones that I described above, and it's also much easier to recover from: just check out the other branch again and hard-reset it to the rebased one.

[1] https://public-inbox.org/git/354f9fed-567f-42c8-9da9-148a5e223022@haller-berlin.de/

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go generate ./...`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [ ] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [ ] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

### CODE REVIEW

**Comment by stefanhaller** [comment]
The test failures with git 2.20 are because of a bug in our todo parser. See https://github.com/fsmiamoto/git-todo-parser/pull/7 for a fix for that.

File: pkg/app/daemon/daemon.go [hunk 1]
```diff
@@ -157,6 +159,26 @@ func NewExitImmediatelyInstruction() Instruction {
 	return &ExitImmediatelyInstruction{}
 }
 
+type FixTodosFileInstruction struct{}
```
Comment (jesseduffield) [inline pkg/app/daemon/daemon.go:162]: I'd rather call this what it is i.e. `RemoveUpdateRefsForCopiedBranch`. I don't consider the default behaviour broken and in need of fixing; rather I consider it a default that we don't like.

**Review by jesseduffield — CHANGES_REQUESTED** [review]
I've read through the messages on the git mailing list and admit I'm conflicted: I also have a strong bias towards making behaviour easy to explain over covering all edge cases, and I have an additional bias of just doing what the git CLI does so that git CLI users don't get surprised when using lazygit.

With that said, I share your experience in that I don't want a copied branch to be treated as a stacked branch and I assume that few people do.

What is good about this PR is that it will be trivially easy to skip this logic and honour the git CLI's default behaviour if there are any users who take issue with the change. If (and when) that happens, I propose we add a config option for it and default to doing things the git CLI way. Until then, let's see if anybody complains.

As for the code, it all looks good, and good work on the tests. Just one comment about naming.

**Comment by stefanhaller** [comment]
@jesseduffield How do you feel about switching our git-todo-parser dependency over to my fork? https://github.com/fsmiamoto/git-todo-parser/pull/7 is a blocker for this PR, but Flavio is sometimes slow at approving PRs, and I tend to make more changes than Flavio in that repo anyway. (I am a collaborator in his repo, but it seems I can't merge my own PRs when they need approval.)

**Comment by jesseduffield** [comment]
@stefanhaller go for it

File: pkg/app/daemon/daemon.go (hunk 1 above)
Comment (stefanhaller) [reply pkg/app/daemon/daemon.go]: Fine with me, changed in f9a285fb10e2 (along with other related symbols).

**Comment by stefanhaller** [comment]
@jesseduffield I added one more commit at the beginning of the branch (920b44811206), please have a look.

**Review by jesseduffield — APPROVED** [review]
Great work!

**Pull request merged**